src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -78,20 +78,13 @@ private void StopRaisingEvents()
                 {
                     // Stop the RunLoop and wait for the thread to exit gracefully
                     Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
-                    _watcherThread.Join();
-                    _watcherRunLoop = IntPtr.Zero;
                 }
 
-                // Clean up the EventStream, if it exists
-                if (!_eventStream.IsInvalid)
-                {
-                    _eventStream = new SafeEventStreamHandle(IntPtr.Zero);
-                }             
-
-                // Cleanup the callback
-                if (_callback != null)
+                // Gurd against us from calling Join on our own background thread (for 
+                // example, calling Dispose or Stop from a callback).
+                if (Thread.CurrentThread != _watcherThread)
                 {
-                    _callback = null;
+                    _watcherThread.Join();
```

Comment:
Thanks.  Though by removing it, please also vet the rest of the implementation to make sure everything will remain consistent if, for example, code quickly starts, then stops, then starts the watcher again.  That needs to be safe to do, even though it could end up with a background processing loop running concurrently with either the code that starts/stops and/or another background processing loop.  This is safe on Windows and Linux because of their implementations... maybe that's what you were joining, because it's not safe?  In which case we either need to join or we need to make sure it's safe (ideally the latter).